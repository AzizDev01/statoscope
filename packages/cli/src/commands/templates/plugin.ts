import { concatTemplateParts, templatePartsByOptions } from './templateParts';
import { FileExt, RuleTemplatesPartsFromFileExt, Template } from './types';

export const pluginTemplate: Template = (templateOptions) => {
  const plugin = templatePartsByOptions(templateOptions, {
    js: jsPlugin,
    ts: tsPlugin,
  });

  return concatTemplateParts(plugin);
};

const jsPlugin: RuleTemplatesPartsFromFileExt[FileExt.js] = {
  imports: {
    commonjs: `const asyncChunkRule = require('./rule.statoscope.js')`,
    esm: `import asyncChunkRule from './rule.statoscope.js';`,
  },
  body: `/**
  * @typedef {import('@statoscope/webpack-model').Prepared} Prepared
  * @returns {import('@statoscope/stats-validator/dist/plugin').PluginDescription<Prepared>} PluginDescription
 */
const newStatoscopePlugin = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: {
    commonjs: 'module.exports = newStatoscopePlugin',
    esm: 'export default newStatoscopePlugin',
  },
};

const tsPlugin: RuleTemplatesPartsFromFileExt[FileExt.ts] = {
  imports: `import type { PluginFn } from '@statoscope/stats-validator/dist/plugin';
import type { Prepared } from '@statoscope/webpack-model';
import asyncChunkRule from './rule.statoscope.ts';`,
  body: `
const newStatoscopePlugin: PluginFn<Prepared> = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: 'export default newStatoscopePlugin',
};
