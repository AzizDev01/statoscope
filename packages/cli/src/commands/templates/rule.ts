export function ruleTemplate(): string {
  return `import { Rule } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';

//Statoscope Plugins examples:
https://github.com/statoscope/statoscope/tree/c9d0c148c1e38a87696d79485f37e598110db80c/packages/stats-validator-plugin-webpack#rules


${asyncChunkRule}

export default asyncChunkRule;`;
}

export const asyncChunkRule = `
type WebpackRule<TParams> = Rule<TParams, Prepared>;

type Params = string[];

const asyncChunkRule: WebpackRule<Params> = (ruleParams, data, api): void => {
  const normalizedParams: string[] = ruleParams ? ruleParams : [];
  const chunks = Array.from(new Set(...data.compilations.map((c) => c.chunks)));
  chunks
    .filter((c) =>
      c.names.some((n) => normalizedParams.includes(n) && c.initial !== false)
    )
    .forEach((f) => {
      api.message(f.names[0] + ' is not async chunk.');
    });
};`;
