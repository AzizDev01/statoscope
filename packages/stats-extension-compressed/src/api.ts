import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Resource, Size } from './generator';

export type API = (compilationId: string, resourceId: string) => Size | null;

const makeAPI: APIFactory<Format, API> = (source) => {
  const sizeResolvers: Map<string, Resolver<string, Resource>> = new Map();

  for (const compilation of source.payload.compilations) {
    sizeResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.resources, (r) => r.id)
    );
  }

  return (compilationId: string, resourceId: string): Size | null => {
    return sizeResolvers.get(compilationId)?.(resourceId)?.size ?? null;
  };
};

export default makeAPI;
