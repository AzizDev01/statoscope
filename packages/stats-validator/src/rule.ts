import { API } from './api';

export type RuleDataInput<TInput> = { reference?: TInput; input: TInput };

export type Rule<TParams, TInput> = (
  params: TParams | null,
  data: RuleDataInput<TInput>,
  api: API
) => void;
