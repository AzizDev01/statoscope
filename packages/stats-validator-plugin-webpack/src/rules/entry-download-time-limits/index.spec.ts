import { API, makeAPI } from '@statoscope/stats-validator/dist/api';
import plugin from '../..';
import statsV5 from '../../../../../test/bundles/simple/stats-prod.json';
import rule, { Params } from './';

function prepareAPI(ruleParams: Params): API {
  const pluginInstance = plugin();
  const prepared = {
    input: pluginInstance.prepare!([{ name: 'stats.json', data: statsV5 }]),
  };
  const api = makeAPI({ warnAsError: false });

  rule(ruleParams, prepared, api);

  return api;
}

type Case = {
  name: string;
  params: Params;
};

const maxDownloadTimeCases: Case[] = [
  {
    name: 'global',
    params: {
      network: 'Slow',
      global: {
        maxDownloadTime: 1,
      },
    },
  },
  {
    name: 'override global',
    params: {
      network: 'Slow',
      global: {
        maxDownloadTime: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxDownloadTime: 1,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      network: 'Slow',
      global: {
        maxDownloadTime: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      network: 'Slow',
      global: {
        maxDownloadTime: 1,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxDownloadTime: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      network: 'Slow',
      useCompressedSize: false,
      global: {
        maxDownloadTime: 1,
      },
    },
  },
];

const maxInitialDownloadTimeCases: Case[] = [
  {
    name: 'global',
    params: {
      network: 'Slow',
      global: {
        maxInitialDownloadTime: 1,
      },
    },
  },
  {
    name: 'override global',
    params: {
      network: 'Slow',
      global: {
        maxInitialDownloadTime: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxInitialDownloadTime: 1,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      network: 'Slow',
      global: {
        maxInitialDownloadTime: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      network: 'Slow',
      global: {
        maxInitialDownloadTime: 1,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxInitialDownloadTime: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      network: 'Slow',
      useCompressedSize: false,
      global: {
        maxInitialDownloadTime: 1,
      },
    },
  },
];

const excludeCases: Case[] = [
  {
    name: 'entry string',
    params: {
      exclude: ['two'],
      network: 'Slow',
      global: {
        maxDownloadTime: 1,
      },
    },
  },
  {
    name: 'entry regexp',
    params: {
      exclude: [/tw/],
      network: 'Slow',
      global: {
        maxDownloadTime: 1,
      },
    },
  },
  {
    name: 'entry object',
    params: {
      exclude: [{ type: 'entry', name: /tw/ }],
      network: 'Slow',
      global: {
        maxDownloadTime: 1,
      },
    },
  },
  {
    name: 'no exclude',
    params: {
      exclude: [/foooo/],
      network: 'Slow',
      global: {
        maxDownloadTime: 1,
      },
    },
  },
];

const maxAsyncSizeCDownloadTime: Case[] = [
  {
    name: 'global',
    params: {
      network: 'Slow',
      global: {
        maxAsyncDownloadTime: 1,
      },
    },
  },
  {
    name: 'override global',
    params: {
      network: 'Slow',
      global: {
        maxAsyncDownloadTime: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxAsyncDownloadTime: 1,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      network: 'Slow',
      global: {
        maxAsyncDownloadTime: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      network: 'Slow',
      global: {
        maxAsyncDownloadTime: 1,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxAsyncDownloadTime: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      network: 'Slow',
      useCompressedSize: false,
      global: {
        maxAsyncDownloadTime: 1,
      },
    },
  },
];

describe('maxDownloadTime', () => {
  test.each(maxDownloadTimeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('maxInitialDownloadTime', () => {
  test.each(maxInitialDownloadTimeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('maxAsyncDownloadTime', () => {
  test.each(maxAsyncSizeCDownloadTime)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('exclude', () => {
  test.each(excludeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});
