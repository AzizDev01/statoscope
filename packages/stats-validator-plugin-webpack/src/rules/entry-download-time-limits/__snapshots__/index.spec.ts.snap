// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exclude entry object 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`exclude entry regexp 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`exclude entry string 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`exclude no exclude 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of assets is 175 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncDownloadTime global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": 1,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of async assets is 3 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncDownloadTime not match - global 1`] = `Array []`;

exports[`maxAsyncDownloadTime not match - override global 1`] = `Array []`;

exports[`maxAsyncDownloadTime override global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": 1,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of async assets is 3 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncDownloadTime with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.8509842722039473,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 248.85880319695724,
            "entry": "two",
            "initialDownloadTime": 246.00781892475328,
            "rule": Object {
              "maxAsyncDownloadTime": 1,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of async assets is 3 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxDownloadTime global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of assets is 175 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxDownloadTime not match - global 1`] = `Array []`;

exports[`maxDownloadTime not match - override global 1`] = `Array []`;

exports[`maxDownloadTime override global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of assets is 175 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxDownloadTime with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 16.041805869654606,
            "entry": "one",
            "initialDownloadTime": 16.041805869654606,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of assets is 16 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.8509842722039473,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 248.85880319695724,
            "entry": "two",
            "initialDownloadTime": 246.00781892475328,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": 1,
              "maxInitialDownloadTime": null,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of assets is 249 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialDownloadTime global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": 1,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of initial assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": 1,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of initial assets is 173 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialDownloadTime not match - global 1`] = `Array []`;

exports[`maxInitialDownloadTime not match - override global 1`] = `Array []`;

exports[`maxInitialDownloadTime override global 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 9.014731959292764,
            "entry": "one",
            "initialDownloadTime": 9.014731959292764,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": 1,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of initial assets is 9 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.71044279399671,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 175.4559968647204,
            "entry": "two",
            "initialDownloadTime": 172.74555407072367,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": 1,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of initial assets is 173 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialDownloadTime with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 0,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 16.041805869654606,
            "entry": "one",
            "initialDownloadTime": 16.041805869654606,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": 1,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download time of initial assets is 16 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "718159f8e48dcecd92c9",
    "details": Array [
      Object {
        "filename": "input.json",
        "payload": Object {
          "context": Object {
            "asyncDownloadTime": 2.8509842722039473,
            "compilation": "718159f8e48dcecd92c9",
            "downloadTime": 248.85880319695724,
            "entry": "two",
            "initialDownloadTime": 246.00781892475328,
            "rule": Object {
              "maxAsyncDownloadTime": null,
              "maxDownloadTime": null,
              "maxInitialDownloadTime": 1,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              downloadTime: #.downloadTime,
              initialDownloadTime: #.initialDownloadTime,
              asyncDownloadTime: #.asyncDownloadTime,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download time of initial assets is 246 ms. It's over the 1 ms limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;
